Source/wayland/WaylandServer+Cursor.m

```diff
@@ -26,59 +26,140 @@
 */
 
 #include "wayland/WaylandServer.h"
-#include <AppKit/NSEvent.h>
-#include <AppKit/NSView.h>
-#include <AppKit/NSWindow.h>
-#include <AppKit/GSWindowDecorationView.h>
-#include <AppKit/GSTheme.h>
-#include <AppKit/NSApplication.h>
+#include "cairo/WaylandCairoShmSurface.h"
+#import <AppKit/NSEvent.h>
+#import <AppKit/NSView.h>
+#import <AppKit/NSWindow.h>
+#import <AppKit/NSCursor.h>
+#import <AppKit/NSGraphics.h>
+#import <AppKit/NSBitmapImageRep.h>
+#import <GNUstepGUI/GSWindowDecorationView.h>
+#import <GNUstepGUI/GSTheme.h>
 #include <linux/input.h>
+#include "wayland-cursor.h"
+
+// XXX should this be configurable by the user?
+#define DOUBLECLICK_DELAY 300
+#define DOUBLECLICK_MOVE_THREASHOLD 3
 
 static void
 pointer_handle_enter(void *data, struct wl_pointer *pointer, uint32_t serial,
 		     struct wl_surface *surface, wl_fixed_t sx_w,
 		     wl_fixed_t sy_w)
 {
-  if (!surface)
-    {
-      NSDebugLog(@"no surface");
-      return;
-    }
-
   WaylandConfig *wlconfig = data;
+
   struct window *window = wl_surface_get_user_data(surface);
+
+  if(window->ignoreMouse)
+  {
+    return;
+  }
+
+  wlconfig->pointer.focus = window;
+
+  if(wlconfig->pointer.captured)
+  {
+    return;
+  }
+
+  [(WaylandServer *)GSCurrentServer() initializeMouseIfRequired];
+
+
+  NSDebugLog(@"[%d] pointer_handle_enter",window->window_id);
+
   float		 sx = wl_fixed_to_double(sx_w);
   float		 sy = wl_fixed_to_double(sy_w);
-  [GSCurrentServer() initializeMouseIfRequired];
+
+
+  if (window && wlconfig->pointer.serial)
+    {
+      NSEvent	      *event;
+      NSPoint		 eventLocation;
+      NSGraphicsContext *gcontext;
+
+      float		 deltaX = sx - wlconfig->pointer.x;
+      float		 deltaY = sy - wlconfig->pointer.y;
+
+      gcontext = GSCurrentContext();
+      eventLocation = NSMakePoint(sx, window->height - sy);
+
+      event = [NSEvent mouseEventWithType:NSMouseEntered
+				 location:eventLocation
+			    modifierFlags:wlconfig->modifiers
+				timestamp:wlconfig->pointer.last_timestamp
+			     windowNumber:window->window_id
+				  context:gcontext
+			      eventNumber:serial
+			       clickCount:0
+				 pressure:0.0
+			     buttonNumber:0
+				   deltaX:deltaX
+				   deltaY:deltaY
+				   deltaZ:0.];
+
+      [GSCurrentServer() postEvent:event atStart:NO];
+    }
 
   wlconfig->pointer.x = sx;
   wlconfig->pointer.y = sy;
-  wlconfig->pointer.focus = window;
-
-  // FIXME: Send NSMouseEntered event.
+  wlconfig->pointer.serial = serial;
+  wlconfig->event_serial = serial;
 }
 
 static void
 pointer_handle_leave(void *data, struct wl_pointer *pointer, uint32_t serial,
 		     struct wl_surface *surface)
 {
-  if (!surface)
+  WaylandConfig *wlconfig = data;
+
+  struct window *window = wl_surface_get_user_data(surface);
+
+  if(window->ignoreMouse)
```

Comment:
It would be great if you could add a space here (between "if" and "(") and in a lot of other places too.
Just try to follow the GNUstep coding conventions.

Reply:
i added some formatting